### ISSUE TITLE
Very slow workflow tasks page for non-Administrators epersons

### ISSUE THREAD

**Issue by peter975** [DSpace/DSpace#9906]
## Describe the bug
DSpace 7.6.2

We've been testing DSpace 7.6.2 and it has become apparent that membership (direct or indirect) of any collection Editors group substantially increases the time it takes to show a list of workflow tasks for an eperson.

There are also large disparities how long workflow tasks take to list between epersons that are Administrators (members of the Administrators group) and non-administrator epersons that are members of a single collection group.

For context, the instance of DSpace we are testing on has ~686 items in workflow and ~900 in workspace.

Our testing has shown:

1. An eperson (non-Administrator) that is a direct member of the Editors group in only a single collection can expect to see workflow tasks listed in ~18 seconds.
2. Adding the same eperson to the Editors role on >1 collections does not seem to appreciably increase the time to list workflow tasks.
3. A member of the Administrators group that is also a member of the samed Editors group in a single collection can expect to see workflow tasks listed in ~3 seconds.
4. The above only apply when you results per page is set to 10. For an eperson in a collection editor role, increasing results per page to 100 means the query takes longer than you'd expect. For testing we've set the max query time to 300 seconds and the listing times out and returns an error.
4. Membership of a collection role group in the pre-dspace7 format (eg COLLECTION_638_WORKFLOW_STEP_2) rather than the current format (eg COLLECTION_8221b33d-a09b-486e-89c3-374f99afcbe2_WORKFLOW_ROLE_editor) seems to result in workflow tasks listing in around the same time as a member of the Administrators group - ie much faster than for a non-Administrator. 

The API call that takes the time is /server/api/discover/search/objects?sort=lastModified,DESC&page=0&size=10&configuration=workflow&embed=thumbnail&embed=item%2Fthumbnail&embed=accessStatus

My colleague Agustina (@amgciadev)  has looked at the code and it seems that while the issue was partly addressed here https://github.com/DSpace/DSpace/issues/9164, the PR didn't fully fix things.

The core problem appears to be that the AInprogressItemConverter (used for both WorkspaceItems and WorkflowItems) embeds all data by default. For example, this section of code always embeds the entire SubmissionDefinition in each of these objects:
https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/AInprogressItemConverter.java#L82-L126
 
The submissionDefinition, as well as Sherpa API calls if the step is enabled, are all still happening unnecessarily

## Expected behavior
Workflow tasks for non-Administrator epersons in collection Editor roles (through collection group membership) should list quickly.

## Related work
https://github.com/DSpace/DSpace/issues/9164
https://github.com/DSpace/DSpace/issues/9053

**Comment by tdonohue** [comment]
Might also be related to #7528 (which I just re-discovered while searching around)

Regardless, this needs a volunteer.

**Comment by asharan2buff** [comment]
I can take this up

**Comment by tdonohue** [comment]
@asharan2buff : Thanks for volunteering.  I'll assign this to you.  Let us know if you need early feedback, or else feel free to create a PR when you have something to review.

**Comment by amgciadev** [comment]
@tdonohue thanks for pointing the fix out. Some preliminary testing shows improvement:
This is the API request that is slow: 

`/server/api/discover/search/objects?sort=lastModified,DESC&page=0&size=10&configuration=workflow&embed=thumbnail&embed=item%2Fthumbnail&embed=accessStatus`

Logged in as a content admin (non full admin), member of many groups we see the following timings for the workflow tasks page (`/mydspace?configuration=workflow`):

Pagination 10: loading times as consistently 16-18 seconds
Pagination 100: loading times 2.8 mins (before this fix was more than 5 mins, and it wouldn't finish - we have increased our apache request timeouts from 1 min to 5 mins just to test these)

**Comment by tdonohue** [comment]
@amgciadev : We discussed your comment in today's Developers Meeting.  Thanks for sharing the exact API request that is slow.  I don't yet have a volunteer to help look into the performance of this endpoint, but it sounds like a possible backend performance bug.  It's also possible that the performance bug you are seeing is related to #8197, which has a possible fix in #9760.  If you or your team find time to do so, you could try deploying #9760 in your test environment to see if it helps as well.

**Comment by amgciadev** [comment]
Hi @tdonohue, we have now deployed #9760 in our test 7.6.2, and done some testing: I'm afraid we don't see any difference in loading times of the workflow tasks as per above. However, I can confirm that it solves the issue described there: adding groups to collection administrator resource policies is now much faster.

Re
>UPDATE: You also could consider temporarily turning off the "accessStatus" feature, just to see if that has any major impact on the performance. Set showAccessStatuses: false in your config.prod.yml in the UI configuration. This may have only minimal impact, but it is worth double checking just in case loading those access status labels is taking longer than expected.

The difference in timings is very small, perhaps a second faster.

**Comment by tdonohue** [comment]
Thanks for the additional testing, @amgciadev.  That is helping to narrow things down.  It's not the embedded access statuses, and it's not the Group2GroupCache.

It does sound likely that we have an issue with either loading group memberships or with the `AInprogressItemConverter` as noted in the original description (or both).

@asharan2buff are you still looking into a fix for this (as you mentioned you were [a month ago](https://github.com/DSpace/DSpace/issues/9906#issuecomment-2424716811))?   I need a status check so that we can reassign this as necessary.  If you are still working on it, you are welcome to do so, but it'd be good to understand what directions you are looking into.

(FYI to all... I'll be out of the office for the next week. Will check back in in December)

**Comment by amgciadev** [comment]
Hi @tdonohue, apologies for the delay. This performance issue is causing substantial operational problems in our repository as loading times are really slow. Re:

> It does sound likely that we have an issue with either loading group memberships or with the AInprogressItemConverter as noted in the original description (or both).

It is definitely both, but I would say that more so around loading group memberships. As mentioned in the original ticket there are unresolved issues relating to AInprogressItemConverter, where all the submission information (including unnecessary form validations as @jlipka mentions), as well as also calls to external services such as Sherpa (if the Sherpa step is configured for the submission), is always being loaded during search results. This is completely unnecessary. @jlipka unfortunately is not as simple as disabling those validations because the fillFromModel method is used both when retrieving search results as well as when loading a single submission (this is where is needed) so cannot be disabled there.

I am sharing our current group structure in case it helps further testing / replicating the issues we are seeing:

Our groups structure is comprised of a Main group (repository managers), containing a subgroup for Content Administrators. This subgroup is then assigned to more than 400 collections with both Collection Administrator and Collection Editor roles.

It would be good to confirm whether @asharan2buff is working on this, as otherwise this important issue would be further delayed. I hope that helps, also please do let us know if / how we can further help with this (testing, providing additional information).

**Comment by tdonohue** [comment]
This may be related to the new ticket #10084

**Comment by tdonohue** [comment]
Because this is high priority, I'm reopening this to volunteers.  Needs a volunteer to investigate how to resolve this.

**Comment by bram-atmire** [comment]
## Analysis: Two independent performance problems

After investigating this issue in the context of PR #11646 (which refactors the Solr authorization queries), I wanted to share an analysis of the two separate bottlenecks at play here.

### Problem 1: Slow Solr query construction — addressed by PR #11646

`SolrServiceResourceRestrictionPlugin.additionalSearchParameters()` runs for every Solr query, including workflow searches. It calls `createLocationQueryForAdministrableItems()`, which walks the `resourcepolicy` table — expensive for users who are collection admins/editors on hundreds of collections. PR #11646 replaces this with a fast Solr-based `createLocationQueryForAdministrableDSOs()`, which should eliminate this bottleneck.

Additionally, `SolrServiceWorkspaceWorkflowRestrictionPlugin` (line 87) calls `groupService.allMemberGroupsSet()` to build the `taskfor:` filter, which involves DB queries for group membership. This is less expensive but still contributes.

### Problem 2: Expensive per-result serialization — NOT addressed by #11646

This is likely the bigger contributor to the remaining slowness. The call chain for **each** of the 10 search results:

1. Solr returns `PoolTask`/`ClaimedTask` objects
2. [`PoolTaskConverter.convert()`](https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/PoolTaskConverter.java#L41) calls `converter.toRest(witem, projection)` on the embedded `XmlWorkflowItem`
3. [`WorkflowItemConverter.convert()`](https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/WorkflowItemConverter.java#L36) calls the inherited `fillFromModel()`
4. [`AInprogressItemConverter.fillFromModel()`](https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/AInprogressItemConverter.java#L66) (lines 78–121) does, **for each result**:
   - Loads the full `SubmissionDefinition` via `submissionConfigService.getSubmissionConfigByCollection()`
   - Iterates over every submission section/step
   - Calls `validate()` on each step — e.g. `MetadataValidation` does multiple DB queries per form field
   - Calls `getData()` on each step — e.g. `SherpaPolicyStep` makes **external HTTP API calls** to Sherpa/RoMEO
   - Upload step loads all bundles/bitstreams, DescribeStep loads all metadata

For 10 results with ~5 steps each, that's roughly ~50 `validate()` + ~50 `getData()` calls per page load, including external API calls. This is the dominant source of the slow response times.

### Confirmation: @jlipka's finding

As @jlipka [noted above](https://github.com/DSpace/DSpace/issues/9906#issuecomment-2563127747), removing the `validate()` calls alone dropped request times from ~2s to ~0.5s, confirming this is a major bottleneck.

### Fix direction

The key issue is that `fillFromModel()` is called identically for both search result listings and single-item views. For search results (e.g. the workflow tasks page), the sections data (validations, step data, Sherpa lookups) is unnecessary — it's only needed when viewing/editing a single submission.

A fix would make the section loading in `fillFromModel()` conditional — skip the `validate()` and `getData()` calls when the `WorkflowItem` is being converted as an embedded object inside a search result, rather than as a standalone resource. This could be signaled via the `Projection` or a dedicated flag.

### This is independent of #11646

The serialization bottleneck (Problem 2) can be addressed in a separate PR, developed in parallel with #11646. Together they should fully resolve this issue.
